src/components/Microsoft.DotNet.UpgradeAssistant/Checks/BypassableReadinessCheck.cs

```diff
@@ -0,0 +1,59 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.DotNet.UpgradeAssistant.Checks
+{
+    public abstract class BypassableReadinessCheck : IUpgradeReadyCheck
+    {
+        private readonly ILogger _logger;
+        private readonly UpgradeOptions _upgradeOptions;
+
+        protected BypassableReadinessCheck(ILogger logger, UpgradeOptions upgradeOptions)
+        {
+            _logger = logger ?? throw new ArgumentNullException(nameof(logger));
+            _upgradeOptions = upgradeOptions ?? throw new ArgumentNullException(nameof(upgradeOptions));
+        }
+
+        public abstract string Id { get; }
+
+        public abstract bool IsBypassable { get; }
+
+        public abstract string TechnologyDetected { get; }
+
+        public abstract string SupportLink { get; }
+
+        protected abstract Task<bool> DoesProjectContainTechnologyAsync(IProject project, CancellationToken token);
+
+        public async Task<bool> IsReadyAsync(IProject project, CancellationToken token)
+        {
+            if (project is null)
+            {
+                // this is not a valid scenario
+                // satisfies CA1062
+                return false;
+            }
+
+            if (_upgradeOptions.ReadinessAcknowledgement)
+            {
+                // the user has bypassed this constraint (opted out of this feature)
+                return true;
+            }
+
+            if (await DoesProjectContainTechnologyAsync(project, token).ConfigureAwait(false))
+            {
+                _logger.LogError("Project {Name} uses the {TechnologyDetected} which is not supported.", project.FileInfo, TechnologyDetected);
```

Comment:
Can we make this a single error? At the moment, this approach will show up good in the UI but if we have an alternate UI in the future it might present things differently and having three errors (instead of one) could be confusing.

Reply:
Latest version makes this a single error